tsfresh/feature_extraction/data.py

```diff
@@ -0,0 +1,343 @@
+        """
+
+        _check_nan(df, column_id)
+
+        if value_columns is None:
+            value_columns = [col for col in df.columns if col not in [column_id, column_sort]]
+
+        if len(value_columns) == 0:
+            raise ValueError("You must provide at least one value column")
+
+        _check_nan(df, *value_columns)
+        _check_colname(*value_columns)
+
+        self.value_columns = value_columns
+        self.column_sort = column_sort
+
+        if self.column_sort is not None:
+            _check_nan(df, column_sort)
+            self.df_grouped = df.sort_values([self.column_sort]).groupby([column_id])
+        else:
+            self.df_grouped = df.groupby([column_id])
+
+    def __len__(self):
+        return self.df_grouped.ngroups * len(self.value_columns)
+
+    def slice(self, offset, length=None):
+        group_offset, column_offset = divmod(offset, len(self.value_columns))
+        kinds = self.value_columns[column_offset:]
+        i = 0
+
+        for group_name, group in itertools.islice(self.df_grouped, group_offset, None):
+            for kind in kinds:
+                i += 1
+                if length is not None and i > length:
+                    return
+                else:
+                    yield Timeseries(group_name, kind, group[kind])
+            kinds = self.value_columns
+
+
+class LongTsFrameAdapter(SliceableTsData):
+
+    def __init__(self, df, column_id, column_kind, column_value, column_sort=None):
+        """
+        Adapter for Pandas DataFrames in long format, where different time series for the same id are
+        labeled by column `column_kind`.
+
+        :param df: the data frame
+        :type df: pd.DataFrame
+
+        :param column_id: the name of the column containing time series group ids
+        :type column_id: str
+
+        :param column_kind: the name of the column containing time series kinds for each id
+        :type column_kind: str
+
+        :param column_value: the name of the column containing time series values
+        :type column_value: str
+
+        :param column_sort: the name of the column to sort on
+        :type column_sort: str|None
+        """
+
+        _check_nan(df, column_id, column_kind, column_value)
+        _check_colname(column_kind)
+        if column_sort is not None:
+            _check_nan(df, column_sort)
+
+        self.column_value = column_value
+        self.column_sort = column_sort
+        self.df_grouped = df.groupby([column_id, column_kind])
+
+    def __len__(self):
+        return len(self.df_grouped)
+
+    def slice(self, offset, length=None):
+        length_or_none = None if length is None else offset + length
+
+        for group_key, group in itertools.islice(self.df_grouped, offset, length_or_none):
+
+            if self.column_sort is not None:
+                group = group.sort_values([self.column_sort])
+
+            yield Timeseries(group_key[0], str(group_key[1]), group[self.column_value])
```

Comment:
The same question here: why is this faster? If I use this function as the `__iter__` in combination with the `partition` function in `TsData`, it really looks very similar.

Reply:
I am not sure, if this is faster. As I said earlier in one comment, so far I just optimized WideTsFrame. Long and dict are just first drafts and I still need to look deeper into it. But to clarify, as you focus on speed here: Speed was not my primary focus of this PR. I was just trying to optimize my implementations.

Reply:
For now I am removing the slice partitioning for LongTsFrameAdapter.